packages/engines/src/index.ts

```diff
@@ -13,38 +14,52 @@ export const DEFAULT_CLI_QUERY_ENGINE_BINARY_TYPE = BinaryType.QueryEngineLibrar
  * Checks if the env override `PRISMA_CLI_QUERY_ENGINE_TYPE` is set to `library` or `binary`
  * Otherwise returns the default
  */
-export function getCliQueryEngineBinaryType(): BinaryType {
-  const envCliQueryEngineType = process.env.PRISMA_CLI_QUERY_ENGINE_TYPE
-  if (envCliQueryEngineType) {
-    if (envCliQueryEngineType === 'binary') {
-      return BinaryType.QueryEngineBinary
-    }
-    if (envCliQueryEngineType === 'library') {
-      return BinaryType.QueryEngineLibrary
-    }
+export function getCliQueryEngineBinaryType(clientEngineType = process.env.PRISMA_CLI_QUERY_ENGINE_TYPE): BinaryType {
+  if (clientEngineType === 'binary') {
+    return BinaryType.QueryEngineBinary
   }
   return DEFAULT_CLI_QUERY_ENGINE_BINARY_TYPE
 }
-export async function ensureBinariesExist() {
-  const binaryDir = path.join(__dirname, '../')
-  let binaryTargets: string[] | undefined
-  if (process.env.PRISMA_CLI_BINARY_TARGETS) {
-    binaryTargets = process.env.PRISMA_CLI_BINARY_TARGETS.split(',')
-  }
 
-  const cliQueryEngineBinaryType = getCliQueryEngineBinaryType()
+type EnsureSomeBinariesExistInput = {
+  clientEngineType: 'library' | 'binary' | 'client'
+  previewFeatures: string[]
+  download: (options: DownloadOptions) => Promise<BinaryPaths>
+}
+
+export async function ensureNeededBinariesExist({
+  clientEngineType,
+  previewFeatures,
+  download,
+}: EnsureSomeBinariesExistInput) {
+  const binaryDir = path.join(__dirname, '../')
 
   const binaries = {
-    [cliQueryEngineBinaryType]: binaryDir,
+    // Schema engine is always needed, as `SchemaEngineWasm` is currently not gated behind a preview feature
     [BinaryType.SchemaEngineBinary]: binaryDir,
   }
+
+  // query engine should only be downloaded if the queryCompiler preview feature is not enabled, or if
+  // QE is enabled explicitly by specifying the engineType generator property.
+  const usesQueryCompiler = previewFeatures.includes('queryCompiler') && clientEngineType === 'client'
```

Comment:
`clientEngineType === 'client'` implies `previewFeatures.includes('queryCompiler')` so it should be enough to just check the `clientEngineType`